rex/joint_pd/joint_pd.py

```diff
@@ -0,0 +1,413 @@
+
+    def save(self, jpd, out_fpath):
+        """
+        Save joint probability distribution to disk
+
+        Parameters
+        ----------
+        jpd : pandas.DataFrame
+            Table of joint probability distribution densities to save
+        out_fpath : str
+            .csv, or .h5 file path to save joint probability
+            distribution to
+        """
+        with self.res_cls(self.res_h5) as f:
+            meta = f['meta', jpd.columns.values]
+
+        logger.info('Writing joint probability distribution to {}'
+                    .format(out_fpath))
+        if out_fpath.endswith('.csv'):
+            jpd.to_csv(out_fpath)
+            meta_fpath = out_fpath.split('.')[0] + '_meta.csv'
+            if os.path.exists(meta_fpath):
+                msg = ("Site meta data already exists at {}!")
+                logger.warning(msg)
+                warn(msg)
+            else:
+                logger.debug('Writing site meta data to {}'
+                             .format(meta_fpath))
+                meta.to_csv(meta_fpath, index=False)
+        elif out_fpath.endswith('.h5'):
+            with h5py.File(out_fpath) as f:
+                f.attrs['rex version'] = __version__
+                for i, n in enumerate(jpd.index.names):
+                    logger.info('')
+                    data = np.array(jpd.index.get_level_values(i))
+                    dset = '{}-bins'.format(n)
+                    logger.debug('Writing {}'.format(dset))
+                    f.create_dataset(dset, shape=data.shape, dtype=data.dtype,
+                                     data=data)
+
+                logger.debug('Writing joint probability density values to jpd')
+                data = jpd.values
+                f.create_dataset('jpd', shape=data.shape, dtype=data.dtype,
+                                 data=data)
+
+                logger.debug('Writing site meta data to meta')
+                meta = to_records_array(meta)
+                f.create_dataset('meta', shape=meta.shape, dtype=meta.dtype,
+                                 data=meta)
+        else:
+            msg = ("Cannot save joint probability distribution, expecting "
+                   ".csv or .h5 path, but got: {}".format(out_fpath))
+            logger.error(msg)
+            raise OSError(msg)
+
+    @classmethod
+    def run(cls, res_h5, dset1, dset2, bins1, bins2,
+            sites=None, res_cls=Resource, hsds=False,
+            max_workers=None, chunks_per_worker=5, out_fpath=None):
+        """
+        Compute joint probability distribution between given datasets using
+        given bins
+
+        Parameters
+        ----------
+        res_h5 : str
+            Path to resource h5 file(s)
+        dset1 : str
+            Dataset 1 to generate joint probability distribution for
+        dset2 : str
+            Dataset 2 to generate joint probabilty distribution for
+        bin1 : tuple
+            (start, stop, step) for dataset 1 bins
```

Comment:
Can't bins also be any of the following options? I find the integer spec to be the most useful. 

https://numpy.org/doc/stable/reference/generated/numpy.histogram2d.html

- If int, the number of bins for the two dimensions (nx=ny=bins).
- If array_like, the bin edges for the two dimensions (x_edges=y_edges=bins).
- If [int, int], the number of bins in each dimension (nx, ny = bins).
- If [array, array], the bin edges in each dimension (x_edges, y_edges = bins).
- A combination [int, array] or [array, int], where int is the number of bins and array is the bin edges.

Reply:
sure!